mex.go

```diff
@@ -43,28 +46,85 @@ const (
 	mexChannelBufferSize = 2
 )
 
+type errNotifier struct {
+	c   chan struct{}
+	err error
+}
+
+func newErrNotifier() errNotifier {
+	return errNotifier{c: make(chan struct{})}
+}
+
+// Notify will store the error and notify all waiters on c that there's an error.
+func (e *errNotifier) Notify(err error) error {
+	// The code should never try to Notify(nil).
+	if err == nil {
+		panic("cannot Notify with no error")
+	}
+
+	// There may be some sort of race where we try to notify the mex twice.
+	select {
+	case <-e.c:
+		return fmt.Errorf("cannot broadcast error: %v, already have: %v", err, e.err)
+	default:
+	}
+
+	e.err = err
+	close(e.c)
+	return nil
+}
+
+// checkErr returns previously notified errors (if any).
+func (e *errNotifier) checkErr() error {
+	select {
+	case <-e.c:
+		return e.err
+	default:
+		return nil
+	}
+}
+
 // A messageExchange tracks this Connections's side of a message exchange with a
 // peer.  Each message exchange has a channel that can be used to receive
 // frames from the peer, and a Context that can controls when the exchange has
 // timed out or been cancelled.
 type messageExchange struct {
 	recvCh    chan *Frame
+	errCh     errNotifier
 	ctx       context.Context
 	msgID     uint32
 	msgType   messageType
 	mexset    *messageExchangeSet
 	framePool FramePool
 
-	// shutdownAtomic is an atomically updated uint32.
+	// The following are atomically updated uint32.
 	shutdownAtomic uint32
+	errChNotified  uint32
+}
+
+// checkError is called before waiting on the mex channels.
+// It returns any existing errors (timeout, cancellation, connection errors).
+func (mex *messageExchange) checkError() error {
+	if err := mex.ctx.Err(); err != nil {
+		return GetContextError(err)
+	}
+
+	return mex.errCh.checkErr()
 }
 
 // forwardPeerFrame forwards a frame from a peer to the message exchange, where
 // it can be pulled by whatever application thread is handling the exchange
 func (mex *messageExchange) forwardPeerFrame(frame *Frame) error {
+	// We want a very specific priority here:
+	// 1. Timeouts/cancellation (mex.ctx errors)
+	// 2. Whether recvCh has buffer space (non-blocking select over mex.recvCh)
+	// 3. Other mex errors (mex.errCh)
+	// Which is why we check the context error only (instead of mex.checkError).
+	// In the mex.errCh case, we do a non-blocking write to recvCh to prioritize it.
 	if err := mex.ctx.Err(); err != nil {
-		return GetContextError(err)
+		return GetContextError(mex.ctx.Err())
```

Comment:
The comment above doesn't make it clear to me why we can't `return GetContextError(err)` here. Is this change required for some reason? The context docs specifically mention that successive calls to `Err()` should return the same value.